src/interchange/src/bin/avro-decode.rs

```diff
@@ -0,0 +1,130 @@
+#![warn(clippy::builtin_type_shadow)]
+#![warn(clippy::duplicate_underscore_argument)]
+#![warn(clippy::double_neg)]
+#![warn(clippy::unnecessary_mut_passed)]
+#![warn(clippy::wildcard_in_or_patterns)]
+#![warn(clippy::collapsible_if)]
+#![warn(clippy::collapsible_else_if)]
+#![warn(clippy::crosspointer_transmute)]
+#![warn(clippy::excessive_precision)]
+#![warn(clippy::overflow_check_conditional)]
+#![warn(clippy::as_conversions)]
+#![warn(clippy::match_overlapping_arm)]
+#![warn(clippy::zero_divided_by_zero)]
+#![warn(clippy::must_use_unit)]
+#![warn(clippy::suspicious_assignment_formatting)]
+#![warn(clippy::suspicious_else_formatting)]
+#![warn(clippy::suspicious_unary_op_formatting)]
+#![warn(clippy::mut_mutex_lock)]
+#![warn(clippy::print_literal)]
+#![warn(clippy::same_item_push)]
+#![warn(clippy::useless_format)]
+#![warn(clippy::write_literal)]
+#![warn(clippy::redundant_closure)]
+#![warn(clippy::redundant_closure_call)]
+#![warn(clippy::unnecessary_lazy_evaluations)]
+#![warn(clippy::partialeq_ne_impl)]
+#![warn(clippy::redundant_field_names)]
+#![warn(clippy::transmutes_expressible_as_ptr_casts)]
+#![warn(clippy::unused_async)]
+#![warn(clippy::disallowed_methods)]
+#![warn(clippy::disallowed_macros)]
+#![warn(clippy::disallowed_types)]
+#![warn(clippy::from_over_into)]
+// END LINT CONFIG
+
+use std::path::PathBuf;
+use std::process;
+
+use anyhow::Context;
+use tokio::fs;
+
+use mz_interchange::avro::Decoder;
+use mz_interchange::confluent;
+use mz_ore::cli;
+use mz_ore::cli::CliConfig;
+
+/// Decode a single Avro row using Materialize's Avro decoder.
+#[derive(clap::Parser)]
+struct Args {
+    /// The path to a file containing the raw bytes of a single Avro datum.
+    data_file: PathBuf,
+    /// The path to a file containing the Avro schema.
+    schema_file: PathBuf,
+    /// Whether the data file uses the Confluent wire format.
+    #[clap(long)]
+    confluent_wire_format: bool,
+}
+
+#[tokio::main]
+async fn main() {
+    let args: Args = cli::parse_args(CliConfig::default());
+    if let Err(e) = run(args).await {
+        println!("{e:#}");
+        process::exit(1);
+    }
+}
+
+async fn run(args: Args) -> Result<(), anyhow::Error> {
+    let mut data = &*fs::read(&args.data_file)
+        .await
+        .context("reading data file")?;
+    if args.confluent_wire_format {
+        let (schema_id, adjusted_data) = confluent::extract_avro_header(data)?;
+        data = adjusted_data;
+        println!("schema id: {schema_id}");
+    }
+    let schema = fs::read_to_string(&args.schema_file)
+        .await
+        .context("reading schema file")?;
+    let ccsr_client: Option<Box<mz_ccsr::Client>> = None;
+    let debug_name = String::from("avro-decode");
+    let confluent_wire_format = false;
+    let mut decoder = Decoder::new(&schema, ccsr_client, debug_name, confluent_wire_format)
+        .context("creating decoder")?;
+    let row = decoder.decode(&mut data).await.context("decoding data")?;
```

Comment:
Why not run this in a loop, so you can have files with multiple records?

Reply:
Didn't have a need for that, and wasn't clear that a bunch of concat'ed Avro datums is a thing—and if so, do you expect the Confluent header on each datum or no?. There's actual Avro data files, which have all sorts of framing etc, which is what avro-tools expects, but what I needed was something that operated on the raw bytes in the simplest way possible.

Reply:
OCF files are a header followed by a series of blocks, each of which contains concatenated records. So, concatenated records are definitely a thing in other formats, and since Avro is self-delimiting nothing stops you from decoding them (IIRC this is why `Decoder` takes a `&mut &[u8]`, so it can read only the first record while advancing the slice to the start of the next one)

On the other hand I do get the idea of doing the simplest possible thing, so the code you have is fine for now

Reply:
Yeah, zero objection to someone improving this in the future based on debugging needs.